snapcraft/internal/meta.py

```diff
@@ -0,0 +1,259 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2016 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import logging
+import re
+import shlex
+import shutil
+import subprocess
+import tempfile
+
+import yaml
+
+from snapcraft import (
+    common,
+)
+
+logger = logging.getLogger(__name__)
+
+
+_MANDATORY_PACKAGE_KEYS = [
+    'name',
+    'version',
+    'description',
+    'summary',
+]
+
+_OPTIONAL_PACKAGE_KEYS = [
+    'architectures',
+    'type',
+    'license-agreement',
+    'license-version',
+    'plugs',
+    'slots',
+]
+
+_KERNEL_KEYS = {
+    'kernel': 'vmlinuz',
+    'initrd': 'initrd.img',
+    'modules': 'lib/modules/{}',
+    'firmware': 'lib/firmware',
+    'dtbs': 'dtbs',
+}
+
+
+class CommandError(Exception):
+    pass
+
+
+def create_snap_packaging(config_data, snap_dir, parts_dir):
+    """Create snap.yaml and related assets in meta.
+    Create  the meta directory and provision it with snap.yaml
+    in the snap dir using information from config_data.
+
+    :param dict config_data: project values defined in snapcraft.yaml.
+    :return: meta_dir.
+    """
+    packaging = _SnapPackaging(config_data, snap_dir, parts_dir)
+    packaging.write_snap_yaml()
+    packaging.setup_assets()
+
+    return packaging.meta_dir
+
+
+class _SnapPackaging:
```

Comment:
I don't need to pass around `snap_dir` and `parts_dir` which if `common.get` weren't used all over would really complicate the API.